### PR TITLE
CASSANDRA-21111 Fix overflowing of SSTable generation id by moving to long instead of integer

### PR DESCRIPTION
**Pull request by smiklosovic**
… integer.

patch by Stefan Miklosovic; reviewed by TBD for CASSANDRA-21111

Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/io/sstable/SequenceBasedSSTableId.java [hunk 1]
```diff
@@ -115,20 +115,20 @@ public boolean isUniqueIdentifier(String str)
         @Override
         public boolean isUniqueIdentifier(ByteBuffer bytes)
         {
-            return bytes != null && bytes.remaining() == Integer.BYTES && bytes.getInt(0) >= 0;
+            return bytes != null && bytes.remaining() == Long.BYTES && bytes.getLong(0) >= 0;
```
Comment (jyothsnakonisa) [inline src/java/org/apache/cassandra/io/sstable/SequenceBasedSSTableId.java:118]: When older SStables are read with this new version of cassandra with these changes, `bytes.remaining() == Long.BYTES` will be false and this method will return false. 

You might want to think about backward compatibility issues, also it would be nice if you can add tests around those scenarios.

File: src/java/org/apache/cassandra/io/sstable/AbstractSSTableSimpleWriter.java [hunk 2]
```diff
@@ -147,13 +147,13 @@ protected SSTableTxnWriter createWriter(SSTable.Owner owner) throws IOException
...
 
-    private static Descriptor createDescriptor(File directory, final String keyspace, final String columnFamily, final SSTableFormat<?, ?> fmt) throws IOException
+    private Descriptor createDescriptor(File directory, final String keyspace, final String columnFamily, final SSTableFormat<?, ?> fmt) throws IOException
     {
         SSTableId nextGen = getNextId(directory, columnFamily);
         return new Descriptor(directory, keyspace, columnFamily, nextGen, fmt);
     }
 
-    private static SSTableId getNextId(File directory, final String columnFamily) throws IOException
+    private SSTableId getNextId(File directory, final String columnFamily) throws IOException
```
Comment (jyothsnakonisa) [inline src/java/org/apache/cassandra/io/sstable/AbstractSSTableSimpleWriter.java:156]: Not a change that you made but In this method, can you please avoid doing `try (Stream<Path> existingPaths = Files.list(directory.toPath()))` in the while loop for CAS to avoid doing full directory scan on each retry for CAS?

File: src/java/org/apache/cassandra/io/sstable/AbstractSSTableSimpleWriter.java [hunk 3]
```diff
@@ -53,7 +53,7 @@ public abstract class AbstractSSTableSimpleWriter implements Closeable
     protected final TableMetadataRef metadata;
     protected final RegularAndStaticColumns columns;
     protected SSTableFormat<?, ?> format = DatabaseDescriptor.getSelectedSSTableFormat();
-    protected static final AtomicReference<SSTableId> id = new AtomicReference<>(SSTableIdFactory.instance.defaultBuilder().generator(Stream.empty()).get());
+    protected final AtomicReference<SSTableId> id = new AtomicReference<>(SSTableIdFactory.instance.defaultBuilder().generator(Stream.empty()).get());
```
Comment (jyothsnakonisa) [inline src/java/org/apache/cassandra/io/sstable/AbstractSSTableSimpleWriter.java:56]: Since `id` is made non static, can you add documentation about having one writer per one sstable? also, if that is the case, this class might be used only by one thread and you might not need thread safety related stuff here.

**Review by jyothsnakonisa — COMMENTED** [review]
Left some comments, can you please add tests for upgrade scenarios for backward compatibility of Ids from int to long and also some tests for checking overflow scenarios

**Comment by smiklosovic** [comment]
closing for reasons in the ticket
